naga/src/front/wgsl/parse/mod.rs

```diff
@@ -1967,9 +1965,9 @@ impl Parser {
                         let mut body = ast::Block::default();
                         if !lexer.skip(Token::Separator(';')) {
                             let (condition, span) = lexer.capture_span(|lexer| {
-                                let condition = self.general_expression(lexer, ctx)?;
-                                lexer.expect(Token::Separator(';'))?;
-                                Ok(condition)
+                                self.general_expression(lexer, ctx).and_then(|condition| {
+                                    lexer.expect(Token::Separator(';')).map(|()| condition)
+                                })
```

Comment:
The original code seems much clearer to me than the `and_then` / `map` version.

Reply:
Maybe I'm barking up the wrong tree, but: the classic structured control flow syntax, with blocks of statements run in sequence, executed for their side effects, has a very fine theoretical pedigree; see Dijkstra's "A Discipline Of Programming", for example, which uses structured control flow as the basis for writing proofs about programs' behavior.

There is no deep reason to prefer 'functional' means like `and_then` and `map` to ordinary control. They are not inherently "cleaner" or "more well-defined". The choice should be made simply on the basis of which is quicker to read, and less distracting to the reader.

To me, the original code - a series of statements, some of which have `?` operators in them - is much more straightforward.

Reply:
The reason this "needed" to change was because I wanted to avoid specifying the error type for the closure, and this let me do it. With such a vociferous complaint, however, I see no reason not to just eat the return type. 😅

Fixed in the force-pushed 5fad2eff.